src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -0,0 +1,260 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    internal static class ClsidMap
+    {
+        private struct ClsidEntry
+        {
+            public string type;
+            public string assembly;
+        }
+
+        public static void Create(MetadataReader metadataReader, string clsidMapPath)
+        {
+            Dictionary<string, ClsidEntry> clsidMap = new Dictionary<string, ClsidEntry>();
+
+            string assemblyName = GetAssemblyName(metadataReader).FullName;
+
+            bool isAssemblyComVisible = IsComVisible(metadataReader, metadataReader.GetAssemblyDefinition());
+
+            foreach (TypeDefinitionHandle type in metadataReader.TypeDefinitions)
+            {
+                TypeDefinition definition = metadataReader.GetTypeDefinition(type);
+
+                if (TypeIsPublic(metadataReader, definition) && TypeIsClass(metadataReader, definition) && IsComVisible(metadataReader, definition, isAssemblyComVisible))
+                {
+                    string guid = GetTypeGuid(metadataReader, definition);
+
```

Comment:
That's one of the changes we discussed offline. We've decided to require users to explicitly specify a GUID for any type they want to expose via COM.

I can't think of any case for a user to have `[ComVisible(true)]` on a class and not specify an explicit GUID if the above statement still holds true. The other option (again assuming the statement above is still valid) would be to silently not expose the type to COM, which would result in a lot of unexpected behavior for users (and probably a lot of duplicate bug reports for us).

Reply:
>I can't think of any case for a user to have [ComVisible(true)] on a class and not specify an explicit GUID if the above statement still holds true.

That happens all the time and in fact is the normal case in .NET Framework. I think @jeffschwMSFT's point here is we want an explicit error or perhaps a wanring about what happened because this requirement will be new to users.

Reply:
This is giving an explicit error to the user. It will be (unless another PR takes the code before this is merged) error NETSDK1086. There error message is in Strings.resx. Do we want to downgrade this to a warning?

Reply:
This is only considering the source-built binary. If we wanted to consider generating a comhost for a previously built binary, that would be done differently (and probably not as an MSBuild integration).

Reply:
> This is only considering the source-built binary

This is good.  Then my only ask is to do a little more due diligence on the likelihood of a case where ComVisible is present, but was benign but is now an error.

Reply:
As a note, this task (and this error) will only execute on projects with the MSBuild `UseComHost` property set. Since this property currently doesn't exist, users will have to opt-in to this behavior.